include/range/v3/action/reverse.hpp

```diff
@@ -0,0 +1,78 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Eric Niebler 2013-2014
+//  Copyright Gonzalo Brito Gadeschi 2017
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_ACTION_REVERSE_HPP
+#define RANGES_V3_ACTION_REVERSE_HPP
+
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/range_traits.hpp>
+#include <range/v3/action/action.hpp>
+#include <range/v3/algorithm/reverse.hpp>
+#include <range/v3/utility/iterator_concepts.hpp>
+#include <range/v3/utility/static_const.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-actions
+        /// @{
+        namespace action
+        {
+            struct reverse_fn
+            {
+            private:
+                friend action_access;
+
+            public:
+                template<typename Rng, typename I = iterator_t<Rng>,
+                    CONCEPT_REQUIRES_(BidirectionalRange<Rng>() and Permutable<I>())>
+                Rng operator()(Rng && rng) const
+                {
+                    ranges::reverse(rng);
+                    return std::forward<Rng>(rng);
+                }
+
+            #ifndef RANGES_DOXYGEN_INVOKED
+                template<typename Rng,
+                    CONCEPT_REQUIRES_(!BidirectionalRange<Rng>())>
+                void operator()(Rng &&) const
+                {
+                    CONCEPT_ASSERT_MSG(BidirectionalRange<Rng>(),
+                        "The object on which action::reverse operates must be a model of the "
+                        "BidirectionalRange concept.");
+                }
+                template<typename Rng, typename I = iterator_t<Rng>,
+                    CONCEPT_REQUIRES_(BidirectionalRange<Rng>() and !Permutable<I>())>
+                void operator()(Rng &&) const
+                {
+                    CONCEPT_ASSERT_MSG(Permutable<I>(),
+                        "The iterator type of the range passed to action::reverse must allow its "
+                        "elements to be permuted; that is, the values must be movable and the "
+                        "iterator must be mutable.");
+              }
```

Comment:
I have typically put all such `CONCEPT_ASSERT`s in a single overload to keep boilerplate under control. I would prefer that here.

Also, that trailing `}` looks funny hanging out in space there.